### ISSUE TITLE
Enhancement: Have a way to run SQLFluff for a git branch, filtering the output to new or modified areas

### ISSUE THREAD

**Issue by barrywhart** [alanmcruickshank/sqlfluff#85]
There is a Python package called [`diff-cover`](https://github.com/Bachmann1234/diff_cover) which provides two command-line tools with similar functionality for Python code:
* `diff-cover` for coverage information
* `diff-quality` for code quality

In both cases, the tool runs the underlying command (e.g. pylint, flake8, or coverage) and filters the results to display *only* the output relating to new or modified parts of the code.

It would be great to have a way to do the same with SQLFluff. `diff-cover` is mentioned only as a similar tool; it may not be directly useful or relevant to addressing this need.

**Comment by barrywhart** [comment]
`diff-cover` is still actively developed (8 releases in 2019). I'll take a look at the implementation or for other tools with similar capabilities. Mostly I think it is leveraging underlying git functionality, then parsing the output -- it's probably pretty mechanical and straightforward, but very useful.

**Comment by barrywhart** [comment]
I found an existing [`diff-cover` issue](https://github.com/Bachmann1234/diff_cover/issues/8) discussing the addition of support for other tools and other languages than Python. I [commented on it](https://github.com/Bachmann1234/diff_cover/issues/8#issuecomment-562785411), asking if they'd consider a PR adding SQLFluff support.

**Comment by barrywhart** [comment]
Some notes on the design of the `diff-quality` tool, specifically how it supports multiple underlying quality checker tools.
* The class [`QualityReporter`](https://github.com/Bachmann1234/diff_cover/blob/1ad7346982124272d081ef838d2b87589c9e45ca/diff_cover/violationsreporters/base.py#L110) has a constructor parameter called `driver`, which interfaces the core diff quality filtering engine with the specific back-end tool, e.g. PyLint, PyFlakes.
* The dictionary of currently supported drivers can be seen [here](https://github.com/Bachmann1234/diff_cover/blob/master/diff_cover/diff_quality_tool.py#L36L49).
* This [module](https://github.com/Bachmann1234/diff_cover/blob/master/diff_cover/violationsreporters/violations_reporter.py) implements most of the drivers. Some have custom code, but many are based on a generic class `RegexBasedDriver`.


To simplify the integration, I suggest adding a new SQLFluff option that, when enabled, displays the quality output in an alternate format that allows use of `RegexBasedDriver`.  Under this option, each line of output would contain:
* Source filename
* Line number
* Character position
* Quality message

For example, rather than this output:

```
== [sql/benchmark_summary_queries/constraints/_postcondition_number_distinct_labels.sql] FAIL
L:   8 | P:   1 | ???? | Found unparsable segment @L008P001: u'FROM\n    `benchmark_...'
L:  15 | P:  32 | ???? | Found unparsable segment @L015P032: u'IS NULL\n    AND exis...'
[sql/benchmark_summary_queries/raw_benchmark_summaries[business_type,size_label,gender_label,age_label].sql] FAIL
L:  18 | P:   9 | ???? | Found unparsable segment @L018P009: u'FROM\n            `be...'
L:  31 | P:  13 | ???? | Found unparsable segment @L031P013: u'SAFE_CAST(NULL AS ST...'
```

the alternate output would look like:

```
[sql/benchmark_summary_queries/constraints/_postcondition_number_distinct_labels.sql] L:   8 | P:   1 | ???? | Found unparsable segment @L008P001: u'FROM\n    `benchmark_...'
[sql/benchmark_summary_queries/constraints/_postcondition_number_distinct_labels.sql] L:  15 | P:  32 | ???? | Found unparsable segment @L015P032: u'IS NULL\n    AND exis...'
[sql/benchmark_summary_queries/raw_benchmark_summaries[business_type,size_label,gender_label,age_label].sql] L:  18 | P:   9 | ???? | Found unparsable segment @L018P009: u'FROM\n            `be...'
[sql/benchmark_summary_queries/raw_benchmark_summaries[business_type,size_label,gender_label,age_label].sql] L:  31 | P:  13 | ???? | Found unparsable segment @L031P013: u'SAFE_CAST(NULL AS ST...'
```

**Comment by barrywhart** [comment]
The maintainer of `diff_cover` replied:
>Sure. Send a pr and I'll add it.

I will take this issue if that's okay. Currently, all their integrations are "direct", i.e. the package directly contains code to integrate with the tool (e.g. PyLint). They mention on the issue that ideally this would be done by adding a plugin system. I plan to look at this as well. There is a package called [`pluggy`](https://pypi.org/project/pluggy/) which I have used in the past and makes this pretty easy.

**Comment by alanmcruickshank** [comment]
>  I suggest adding a new SQLFluff option that, when enabled, displays the quality output in an alternate format

I like this option. 👆. I can imagine over time that we might have several different output formats for different use cases, and that the formatting and output should be abstracted away from the linter itself.

That would require:
1. Having a `--format` option as one of the common options for all cli commands, which is also available in the `config` library (which would enable it to be set in config files).
2. Some surgery on all the output bits - ideally as part of the `cli.formatters` module so that all the format specific output gets handled there.

In particular for step 2, the code is a bit tangled around this. Feel free to attempt some fairly drastic surgery to separate **linting logic** from **formatting logic**.

**Comment by barrywhart** [comment]
After posting my earlier comment, I ended up taking a different approach: On the `diff_cover` project, they had discussed letting new tools like SQLFluff integrate as *plugins* rather than having their tool invoke other tools as a separate process.

The plugin approach lets SQLFluff provide its own code that calls directly into the underlying Linter class. See the `SQLFluffViolationReporter` class in the PR.

Please have a look at the PR and let me know what you think.

**Comment by alanmcruickshank** [comment]
@barrywhart - is this issue now closed? I think based on #128 I assume that it is.

**Comment by barrywhart** [comment]
Yes. I'll close it.
